File: lib/constructs/code-assets/python/ddb.py
import json
import os
import boto3
import uuid
from datetime import datetime
from botocore.exceptions import ClientError

# Initialize DynamoDB client
dynamodb = boto3.resource('dynamodb', region_name=os.environ['REGION'])
table = dynamodb.Table(os.environ['TABLE_NAME'])

def handler(event, context):
    try:
        # Get the request body
        body = event.get('body', '{}')
        if isinstance(body, str):
            body = json.loads(body)

        # Generate unique ID and timestamp
        item_id = str(uuid.uuid4())
        current_time = datetime.utcnow().isoformat()

        # Prepare item for DynamoDB
        item = {
            'id': item_id,
            'type': body.get('type', 'default_type1'),
            'data': body.get('data', {}),
            'createdAt': current_time,
            'ttl': body.get('ttl'),  # Optional: TTL timestamp in seconds
        }

        # Write to DynamoDB
        table.put_item(Item=item)

        return {
            'statusCode': 200,
            'body': json.dumps({
                'message': 'Item successfully written to DynamoDB',
                'itemId': item_id
            }),
            'headers': {
                'Content-Type': 'application/json'
            }
        }

    except ClientError as e:
        return {
            'statusCode': 500,
            'body': json.dumps({
                'error': f"DynamoDB error: {str(e)}"
            }),
            'headers': {
                'Content-Type': 'application/json'
            }
        }
    except Exception as e:
        return {
            'statusCode': 500,
            'body': json.dumps({
                'error': str(e)
            }),
            'headers': {
                'Content-Type': 'application/json'
            }
        }

File: lib/constructs/code-assets/nodejs/ddb.js
const AWS = require('aws-sdk');
const dynamoDB = new AWS.DynamoDB.DocumentClient();

exports.handler = async (event) => {
    // Configure CloudWatch logging
    console.log('Received event:', JSON.stringify(event, null, 2));

    try {
        // Parse the incoming event body if it exists
        const item = event.body ? JSON.parse(event.body) : event;

        // Add timestamp for record keeping
        item.timestamp = new Date().toISOString();

        // DynamoDB put parameters
        const params = {
            TableName: process.env.TABLE_NAME, // Define this in Lambda environment variables
            Item: item,
            // Optional: Add condition expression to prevent overwriting existing items
            // ConditionExpression: 'attribute_not_exists(id)'
        };

        // Write to DynamoDB
        await dynamoDB.put(params).promise();

        console.log('Successfully wrote item:', JSON.stringify(item, null, 2));

        // Return success response
        return {
            statusCode: 200,
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                message: 'Item successfully created',
                item: item
            })
        };

    } catch (error) {
        console.error('Error:', error);

        // Return error response
        return {
            statusCode: error.statusCode || 500,
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                message: 'Error creating item',
                errorMessage: error.message,
                errorType: error.name
            })
        };
    }
};
